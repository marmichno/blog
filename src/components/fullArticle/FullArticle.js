import React from 'react';
import { useState, useEffect } from 'react';
import FullArticleCSS from './fullArticle.module.css';
import {useSelector} from 'react-redux';
import {getArticle} from '../../requests/getArticle';
import {ArticleComments} from '../articleComments/ArticleComments';

export const FullArticle = () => {

    const [article, setArticle] = useState([]);
    const selectedArticleId = useSelector(state => state.chooseArticleReducer);
    const [favorite, setFavorite] = useState(false);

    useEffect(() => {
        getArticleRequest();
        checkIfFavorite();
    },[]);

    const checkIfFavorite = () => {
        let favorites = localStorage.getItem('favorites');
        favorites = favorites ? favorites.split(',') : [];

        favorites = favorites.filter(value => parseInt(value) === selectedArticleId ? true : false);

        setFavorite(favorites.length > 0 ? true : false);
    }

    const addToFavorite = () => {
        let favorites = localStorage.getItem('favorites');
        favorites = favorites ? favorites.split(',') : [];
        favorites.push(selectedArticleId);
        localStorage.setItem('favorites', favorites.toString());
        setFavorite(true);
    }

    const getArticleRequest = async () => {
        const response = await getArticle(selectedArticleId);
        setArticle(response);
    }

    if(article !== undefined){
        return(
            <div className={FullArticleCSS.articleContainer}>
                <div className={FullArticleCSS.articleContainer__heroContainer}>
                    <div className={FullArticleCSS.articleContainer__heroContainer__imageContainer}></div>
                    <div className={FullArticleCSS.articleContainer__heroContainer__textContainer}>
                        <h2 className={FullArticleCSS.articleContainer__heroContainer__textContainer__header}>{article.title}</h2>
                        <p className={FullArticleCSS.articleContainer__heroContainer__textContainer__author}>07/12/2021 | John Doe</p>
                        { favorite ?
                        <div className={FullArticleCSS.articleContainer__heroContainer__textContainer__favorite}></div>
                        :
                        null
                    }
                    </div>
                </div>

                <div className={FullArticleCSS.articleContainer__fullTextContainer}>
                    <p className={FullArticleCSS.articleContainer__fullTextContainer__text}>
                        {article.body}
                    </p>
                </div>
                {!favorite ?
                    <div className={FullArticleCSS.articleContainer__addToFavoritesContainer}>
                        <h2 className={FullArticleCSS.articleContainer__addToFavoritesContainer__text}>Did you enjoy reading this article? Add it to 
                            <button className={FullArticleCSS.articleContainer__addToFavoritesContainer__text__button} onClick={addToFavorite}>favorites</button>
                        </h2>
                    </div>
                :
                null
                }
                <ArticleComments/>
            </div>
        )
    }else{
        return(
            <div className={FullArticleCSS.articleContainer}></div>
        )
    }
}