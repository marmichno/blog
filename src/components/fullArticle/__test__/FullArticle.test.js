import { createStore, applyMiddleware} from 'redux';
import * as redux from 'react-redux';
import { Provider } from 'react-redux';
import {FullArticle} from '../FullArticle';
import {fireEvent, render, waitFor} from '@testing-library/react';
import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import validArticle from '../__mocks__/validArticle.json';
import TestRenderer from "react-test-renderer";

const fakeLocalStorage = (() => {
    let store = {};
  
    return {
      getItem: (key) => {
        return store[key] || null;
      },
      setItem: (key, value) => {
        store[key] = value.toString();
      },
      removeItem: (key) => {
        delete store[key];
      },
      clear: () => {
        store = {};
      }
    };
  })();

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(validArticle),
  })
);

beforeAll(() => {
    Object.defineProperty(window, 'localStorage', {
        value: fakeLocalStorage,
    });
});

afterEach(() => {
    jest.clearAllMocks();
});

const store = createStore(() => [], {}, applyMiddleware());

test("Article renders with valid json", async () => {

    const {getAllByText} = render(<Provider store={store}><FullArticle/></Provider>);

    await waitFor(() => {
        expect(getAllByText('est rerum tempore vitae sequi sint nihil reprehenderit dolor beatae ea dolores neque fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis qui aperiam non debitis possimus qui neque nisi nulla'));
    })
});

test("Component renders with api failure", async () => {

    fetch.mockImplementationOnce(() => Promise.reject("API FAILURE"));
    
      const {getByTestId} = render(<Provider store={store}><FullArticle/></Provider>);
    
      await waitFor(() => {
        expect(getByTestId("articleContainer")).toBeTruthy();
      })
    });

test("Favorite button is visible if article is not set to favorite", async () => {

    const {getByTestId} = render(<Provider store={store}><FullArticle/></Provider>);

    await waitFor(() => {
        expect(getByTestId("favoriteButton")).toBeTruthy();
    })
});

test("Favorite button isnt visible after adding to favorites and article is saved in localstorage", async () => {

    const spy = jest.spyOn(redux, 'useSelector');
    spy.mockReturnValue(2);

    const {queryByTestId} = render(<Provider store={store}><FullArticle/></Provider>);

    await waitFor(() => {
        fireEvent.click(queryByTestId("favoriteButton"));
    })

    await waitFor(() => {
        expect(queryByTestId("favoriteButton")).toBeNull();
        expect(window.localStorage.getItem('favorites')).toEqual('2');
    })
});

test("matches snapshot", () => {
    const tree = TestRenderer.create(<Provider store={store}><FullArticle/></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });