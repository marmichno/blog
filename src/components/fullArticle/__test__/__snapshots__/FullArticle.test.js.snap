// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="articleContainer"
  data-testid="articleContainer"
>
  <div
    className="articleContainer__heroContainer"
  >
    <div
      className="articleContainer__heroContainer__imageContainer"
    />
    <div
      className="articleContainer__heroContainer__textContainer"
    >
      <h2
        className="articleContainer__heroContainer__textContainer__header"
      />
      <p
        className="articleContainer__heroContainer__textContainer__author"
      >
        07/12/2021 | John Doe
      </p>
    </div>
  </div>
  <div
    className="articleContainer__fullTextContainer"
  >
    <p
      className="articleContainer__fullTextContainer__text"
    />
  </div>
  <div
    className="articleContainer__addToFavoritesContainer"
  >
    <h2
      className="articleContainer__addToFavoritesContainer__text"
    >
      Did you enjoy reading this article? Add it to
      <button
        className="articleContainer__addToFavoritesContainer__text__button"
        data-testid="favoriteButton"
        onClick={[Function]}
      >
        favorites
      </button>
    </h2>
  </div>
  <div
    className="commentsContainer"
    data-testid="commentsContainer"
  >
    <div
      className="commentsContainer__postCommentContainer"
    >
      <label
        className="commentsContainer__postCommentContainer__label"
        htmlFor="userName"
      >
        <span
          className="commentsContainer__postCommentContainer__span"
        >
          <p
            className="commentsContainer__postCommentContainer__span__paragraph"
          >
            Name
          </p>
        </span>
      </label>
      <input
        autoComplete="off"
        className="commentsContainer__postCommentContainer__input"
        name="userName"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <label
        className="commentsContainer__postCommentContainer__label"
        htmlFor="userEmail"
      >
        <span
          className="commentsContainer__postCommentContainer__span"
        >
          <p
            className="commentsContainer__postCommentContainer__span__paragraph"
          >
            Email
          </p>
        </span>
      </label>
      <input
        autoComplete="off"
        className="commentsContainer__postCommentContainer__input"
        name="userEmail"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <label
        className="commentsContainer__postCommentContainer__label"
        htmlFor="commentBody"
      >
        <span
          className="commentsContainer__postCommentContainer__span"
        >
          <p
            className="commentsContainer__postCommentContainer__span__paragraph"
          >
            Comment
          </p>
        </span>
      </label>
      <textarea
        autoComplete="off"
        className="commentsContainer__postCommentContainer__textArea"
        name="commentBody"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <button
        className="commentsContainer__postCommentContainer__button"
        data-testid="postRequest"
        onClick={[Function]}
      >
        Add comment
      </button>
    </div>
  </div>
</div>
`;
